### PR TITLE
repmatch_gff3 command line standardization

### PR DESCRIPTION
**Pull request by gregvonkuster**
Due to csv module differences between python 2 and 3, this tool currently requires Python 2.  I took a shot at supporting Python 3, but didn't quite get there yet.

### CODE REVIEW

File: tools/repmatch_gff3/repmatch_gff3_macros.xml [hunk 1]
```diff
@@ -3,6 +3,7 @@
     <token name="@WRAPPER_VERSION@">1.0</token>
     <xml name="requirements">
         <requirements>
+            <requirement type="package" version="2.7.13">python</requirement>
```
Comment (bgruening) [inline tools/repmatch_gff3/repmatch_gff3_macros.xml]: This is needed? Is the script not python3 ready?

Comment (gregvonkuster) [inline tools/repmatch_gff3/repmatch_gff3_macros.xml]: Yeah, there's some differences with the csv Python module between 2 and 3.  I took a look at getting things to work, but haven't found fixes for:

- the 2 writer produces ints while the 3 writer produces floats in some cases
- there are mathematical differences in the produced outputs.

I'll keep this on my list to figure out as time permits.  ;)

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 2]
```diff
@@ -23,14 +23,19 @@
...
 class Replicate(object):
 
     def __init__(self, id, dataset_path):
         self.id = id
         self.dataset_path = dataset_path
-        self.parse(csv.reader(open(dataset_path, 'rt'), delimiter='\t'))
+        if ISPY3:
+            fh = open(dataset_path, 'r', newline='')
+        else:
+            fh = open(dataset_path, 'rb', lineterminator='\n')
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: There's no `lineterminator` parameter for [open()](https://docs.python.org/2.7/library/functions.html#open) in Python2, I think you should remove it.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 3]
```diff
@@ -445,7 +455,7 @@ def search_for_matches(group):
                              group.num_replicates,
                              group.median_distance,
                              group.value_sum)
-            for peak in group.peaks.values():
+            for peak in list(group.peaks.values()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 4]
```diff
@@ -39,24 +44,23 @@ def parse(self, reader):
...
             mid = int(mid)
             midplus = int(midplus)
             value = float(value)
-            distance = int(distance)
             if cname not in self.chromosomes:
                 self.chromosomes[cname] = Chromosome(cname)
             chrom = self.chromosomes[cname]
             chrom.add_peak(Peak(cname, mid, value, distance, self))
-        for chrom in self.chromosomes.values():
+        for chrom in list(self.chromosomes.values()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 5]
```diff
@@ -39,24 +44,23 @@ def parse(self, reader):
...
             chrom.sort_by_index()
 
     def filter(self, up_limit, low_limit):
-        for chrom in self.chromosomes.values():
+        for chrom in list(self.chromosomes.values()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 6]
```diff
@@ -39,24 +44,23 @@ def parse(self, reader):
...
             chrom.filter(up_limit, low_limit)
 
     def size(self):
-        return sum([len(c.peaks) for c in self.chromosomes.values()])
+        return sum([len(c.peaks) for c in list(self.chromosomes.values())])
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 7]
```diff
@@ -107,33 +111,33 @@ def add_peak(self, repid, peak):
 
     @property
     def chrom(self):
-        return self.peaks.values()[0].chrom
+        return list(self.peaks.values())[0].chrom
 
     @property
     def midpoint(self):
-        return median([peak.midpoint for peak in self.peaks.values()])
+        return int(median([peak.midpoint for peak in list(self.peaks.values())]))
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: The added `list(...)` is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 8]
```diff
@@ -107,33 +111,33 @@ def add_peak(self, repid, peak):
...
 
     @property
     def num_replicates(self):
         return len(self.peaks)
 
     @property
     def median_distance(self):
-        return median([peak.distance for peak in self.peaks.values()])
+        return int(median([peak.distance for peak in list(self.peaks.values())]))
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: The added `list(...)` is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 9]
```diff
@@ -107,33 +111,33 @@ def add_peak(self, repid, peak):
...
 
     @property
     def value_sum(self):
-        return sum([peak.value for peak in self.peaks.values()])
+        return sum([peak.value for peak in list(self.peaks.values())])
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 10]
```diff
@@ -107,33 +111,33 @@ def add_peak(self, repid, peak):
...
 
     def normalized_value(self, med):
         values = []
-        for peak in self.peaks.values():
+        for peak in list(self.peaks.values()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 11]
```diff
@@ -148,16 +152,16 @@ def add(self, x):
     def graph_series(self):
         x = []
         y = []
-        for key, val in self.dist.items():
+        for key, val in list(self.dist.items()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 12]
```diff
@@ -148,16 +152,16 @@ def add(self, x):
...
             x.append(key)
             y.append(val)
         return x, y
 
     def mode(self):
-        return max(self.dist.items(), key=lambda data: data[1])[0]
+        return max(list(self.dist.items()), key=lambda data: data[1])[0]
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 13]
```diff
@@ -148,16 +152,16 @@ def add(self, x):
...
 
     def size(self):
-        return sum(self.dist.values())
+        return sum(list(self.dist.values()))
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 14]
```diff
@@ -321,7 +325,12 @@ def perform_process(dataset_paths, galaxy_hids, method, distance, step, num_requ
 
     def td_writer(file_path):
         # Returns a tab-delimited writer for a certain output
-        return csv.writer(open(file_path, 'wt'), delimiter='\t')
+        if ISPY3:
+            fh = open(file_path, 'w', newline='')
+            return csv.writer(fh, delimiter='\t', quoting=csv.QUOTE_NONE)
+        else:
+            fh = open(file_path, 'wb', lineterminator='\n')
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: As above, I think you should remove `lineterminator`

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 15]
```diff
@@ -392,9 +399,9 @@ def search_for_matches(group):
                             break
                 # Attempt to enlarge existing peak groups
                 for group in peak_groups:
-                    old_peaks = group.peaks.values()[:]
+                    old_peaks = list(group.peaks.values())[:]
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: You can remove the `[:]` (using `list()` already creates a copy in Python2).

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 16]
```diff
@@ -418,25 +425,28 @@ def search_for_matches(group):
         freq.add(group.num_replicates)
     # Collect together the remaining unmatched_peaks
     for replicate in replicates:
-        for chromosome in replicate.chromosomes.values():
+        for chromosome in list(replicate.chromosomes.values()):
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 17]
```diff
@@ -418,25 +425,28 @@ def search_for_matches(group):
...
             for peak in chromosome.peaks:
                 freq.add(1)
                 unmatched_peaks.append(peak)
     # Average the unmatched_peaks count in the graph by # replicates
-    med = median([peak.value for group in peak_groups for peak in group.peaks.values()])
+    med = median([peak.value for group in peak_groups for peak in list(group.peaks.values())])
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

File: tools/repmatch_gff3/repmatch_gff3_util.py [hunk 18]
```diff
@@ -418,25 +425,28 @@ def search_for_matches(group):
...
     for replicate in replicates:
-        replicate.median = median([peak.value for group in peak_groups for peak in group.peaks.values() if peak.replicate == replicate])
+        replicate.median = median([peak.value for group in peak_groups for peak in list(group.peaks.values()) if peak.replicate == replicate])
```
Comment (nsoranzo) [inline tools/repmatch_gff3/repmatch_gff3_util.py]: This change is not needed.

**Comment by gregvonkuster** [comment]
@nsoranzo Thanks for the review.  So basically no changes are needed for supporting Python3?  I was getting all kiinds of errors.  I'll make all of the changes you suggest though.

**Comment by nsoranzo** [comment]
> So basically no changes are needed for supporting Python3? I was getting all kinds of errors.

Not all the changes suggested by `2to3` are needed, and often other changes (not suggested by `2to3`) are necessary... Human programmers are still needed apparently ;)

**Comment by gregvonkuster** [comment]
Thanks a bunch @nsoranzo ;)

**Pull request merged**